platform/os_windows.go

```diff
@@ -257,43 +248,72 @@ func (p *execClient) ExecutePowershellCommandWithContext(ctx context.Context, co
 	if err != nil {
-		errMsg := fmt.Sprintf("Failed to check the existent of hns state path due to error %s", err.Error())
-		log.Printf(errMsg)
-		return errors.Errorf(errMsg)
+		if errors.Is(err, registry.ErrNotExist) {
+			return nil
+		}
+		return errors.Wrap(err, "could not open registry key")
 	}
-	if strings.EqualFold(exists, "false") {
-		log.Printf("hns state path does not exist, skip setting SdnRemoteArpMacAddress")
-		return nil
+	defer k.Close()
+	// set the key value
+	if err = setSDNRemoteARPMACAddress(k); err != nil {
+		return errors.Wrap(err, "could not set registry key")
 	}
-	if sdnRemoteArpMacAddressSet == false {
-		result, err := execClient.ExecutePowershellCommand(GetSdnRemoteArpMacAddressCommand)
-		if err != nil {
-			return err
-		}
-
-		// Set the reg key if not already set or has incorrect value
-		if result != SDNRemoteArpMacAddress {
-			if _, err = execClient.ExecutePowershellCommand(SetSdnRemoteArpMacAddressCommand); err != nil {
-				log.Printf("Failed to set SDNRemoteArpMacAddress due to error %s", err.Error())
-				return err
-			}
+	// connect to the service manager
+	m, err := mgr.Connect()
+	if err != nil {
+		return errors.Wrap(err, "could not connect to service manager")
+	}
+	defer m.Disconnect() //nolint:errcheck // ignore error
+	// open the HNS service
+	service, err := m.OpenService("hns")
+	if err != nil {
+		return errors.Wrap(err, "could not access service")
+	}
+	defer service.Close()
+	return errors.Wrap(restartService(service), "could not restart service")
+}
 
-			log.Printf("[Azure CNS] SDNRemoteArpMacAddress regKey set successfully. Restarting hns service.")
-			if _, err := execClient.ExecutePowershellCommand(RestartHnsServiceCommand); err != nil {
-				log.Printf("Failed to Restart HNS Service due to error %s", err.Error())
-				return err
-			}
-		}
+type key interface {
+	SetStringValue(name, value string) error
+}
 
-		sdnRemoteArpMacAddressSet = true
+func setSDNRemoteARPMACAddress(k key) error {
+	// Set the reg key
+	// was "Set-ItemProperty -Path HKLM:\\SYSTEM\\CurrentControlSet\\Services\\hns\\State -Name SDNRemoteArpMacAddress -Value \"12-34-56-78-9a-bc\""
+	if err := k.SetStringValue("SDNRemoteArpMacAddress", SDNRemoteArpMacAddress); err != nil {
+		return errors.Wrap(err, "could not set registry key")
 	}
-
+	log.Printf("[Azure CNS] SDNRemoteArpMacAddress regKey set successfully. Restarting hns service.")
 	return nil
 }
 
+type serv interface {
+	Control(code svc.Cmd) (svc.Status, error)
+	Query() (svc.Status, error)
+	Start(...string) error
+}
+
+func restartService(s serv) error {
+	// Stop the service
+	_, err := s.Control(svc.Stop)
+	if err != nil {
+		return errors.Wrap(err, "could not stop service")
+	}
+	// Wait for the service to stop
+	for status, err := s.Query(); status.State != svc.Stopped; status, err = s.Query() {
+		if err != nil {
+			return errors.Wrap(err, "could not query service status")
+		}
+		time.Sleep(500 * time.Millisecond) //nolint:gomnd // 500ms
+	}
```

Comment:
Potentially infinite loop? Probably better to add a timeout via context and use tickers.

Reply:
rewrote with ticker while keeping the do-while semantics